### PR TITLE
Chapter 11: Account Activation

### PR DESCRIPTION
**Pull request by ainulfarizah**
ready

### CODE REVIEW

**Comment by ainulfarizah** [comment]
ready

File: app/mailers/application_mailer.rb [hunk 1]
```diff
@@ -1,4 +1,5 @@
 class ApplicationMailer < ActionMailer::Base
   default from: 'from@example.com'
```
Comment (huongnt-2545) [inline app/mailers/application_mailer.rb]: remove this line

File: config/environments/development.rb [hunk 2]
```diff
@@ -35,6 +50,7 @@
 
   # Don't care if the mailer can't send.
   config.action_mailer.raise_delivery_errors = false
+  host = 'example.com' # Don't use this literally; use your local dev host instead
```
Comment (huongnt-2545) [inline config/environments/development.rb]: remove this line, it's no need

File: config/environments/production.rb [hunk 3]
```diff
@@ -117,4 +117,19 @@
...
+  ActionMailer::Base.smtp_settings = {
+    address: 'smtp.sendgrid.net',
+    port: '587',
+    authentication: :plain,
+    user_name: ENV['SENDGRID_USERNAME'],
+    password: ENV['SENDGRID_PASSWORD'],
+    domain: 'heroku.com',
+    enable_starttls_auto: true
+  }
+end
```
Comment (huongnt-2545) [inline config/environments/production.rb:134]: redundant this line?

File: test/mailers/previews/user_mailer_preview.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
+# Preview all emails at http://localhost:3000/rails/mailers/user_mailer
```
Comment (huongnt-2545) [inline test/mailers/previews/user_mailer_preview.rb]: delete this file

File: test/mailers/user_mailer_test.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
+require "test_helper"
```
Comment (huongnt-2545) [inline test/mailers/user_mailer_test.rb]: delete this file

**Comment by ainulfarizah** [comment]
ready

File: app/controllers/sessions_controller.rb [hunk 6]
```diff
@@ -5,9 +5,14 @@ def new; end
...
-      params[:session][:remember_me] == "1" ? remember(@user) : forget(@user)
-      redirect_back_or @user
+      if @user.activated?
+        log_in @user
+        params[:sessions][:remember_me] == "1" ? remember(@user) : forget(@user)
+        redirect_back_or @user
+      else
+        message = t(:not_activated)
+        flash[:warning] = message
+        redirect_to home_url
```
Comment (huongnt-2545) [inline app/controllers/sessions_controller.rb:15]: you seem missing the `end` keyword for conditional
```
if ...
else
...
end
```

File: app/helpers/sessions_helper.rb [hunk 7]
```diff
@@ -9,9 +9,9 @@ def current_user
       @current_user ||= User.find_by id: user_id
     elsif user_id = cookies.signed[:user_id]
       @user = User.find_by id: user_id
-      if @user&.authenticate(cookies[:remember_token])
+      if @user&.authenticated?(:remember, cookies[:remember_token])
         log_in @user
-        @current_user = @user
+        @current_user = user
```
Comment (huongnt-2545) [inline app/helpers/sessions_helper.rb]:
```suggestion
        @current_user = @user
```

**Comment by ainulfarizah** [comment]
ready

**Comment by ngocvt-0484** [comment]
merged

**Pull request merged**